src/kem.c

```diff
@@ -0,0 +1,393 @@
+/* Copyright (C) 2025 Simo Sorce <simo@redhat.com>
+   SPDX-License-Identifier: Apache-2.0 */
+
+#include "provider.h"
+#include "kem.h"
+#include <string.h>
+#include <openssl/core_dispatch.h>
+#include <openssl/core_names.h>
+#include <openssl/params.h>
+#include <openssl/err.h>
+#include <openssl/proverr.h>
+#include <openssl/evp.h>
+
+struct p11prov_kem_ctx {
+    P11PROV_CTX *provctx;
+    P11PROV_OBJ *key;
+    int op;
+};
+typedef struct p11prov_kem_ctx P11PROV_KEM_CTX;
+
+DISPATCH_MLKEM_FN(newctx);
+DISPATCH_MLKEM_FN(freectx);
+DISPATCH_MLKEM_FN(encapsulate_init);
+DISPATCH_MLKEM_FN(encapsulate);
+DISPATCH_MLKEM_FN(decapsulate_init);
+DISPATCH_MLKEM_FN(decapsulate);
+DISPATCH_MLKEM_FN(set_ctx_params);
+DISPATCH_MLKEM_FN(settable_ctx_params);
+static int p11prov_mlkem_init(void *vctx, int op, void *vkey,
+                              const OSSL_PARAM params[]);
+
+static void *p11prov_mlkem_newctx(void *provctx)
+{
+    P11PROV_CTX *ctx = (P11PROV_CTX *)provctx;
+    P11PROV_KEM_CTX *kemctx;
+
+    kemctx = OPENSSL_zalloc(sizeof(P11PROV_KEM_CTX));
+    if (kemctx == NULL) {
+        return NULL;
+    }
+
+    kemctx->provctx = ctx;
+    return kemctx;
+}
+
+static void p11prov_mlkem_freectx(void *vctx)
+{
+    P11PROV_KEM_CTX *ctx = vctx;
+
+    if (ctx == NULL) {
+        return;
+    }
+
+    p11prov_obj_free(ctx->key);
+    OPENSSL_clear_free(ctx, sizeof(P11PROV_KEM_CTX));
+}
+
+static int p11prov_mlkem_init(void *vctx, int op, void *vkey,
+                              const OSSL_PARAM params[])
+{
+    P11PROV_KEM_CTX *ctx = vctx;
+    P11PROV_OBJ *key = vkey;
+    CK_RV ret;
+
+    if (vctx == NULL || vkey == NULL) {
+        return RET_OSSL_ERR;
+    }
+
+    ret = p11prov_ctx_status(ctx->provctx);
+    if (ret != CKR_OK) {
+        return RET_OSSL_ERR;
+    }
+
+    p11prov_obj_free(ctx->key);
+    ctx->key = p11prov_obj_ref(key);
+    if (ctx->key == NULL) {
+        P11PROV_raise(ctx->provctx, CKR_ARGUMENTS_BAD, "Invalid object");
+        return RET_OSSL_ERR;
+    }
+
+    if (p11prov_obj_get_key_type(ctx->key) != CKK_ML_KEM) {
+        P11PROV_raise(ctx->provctx, CKR_KEY_TYPE_INCONSISTENT,
+                      "Not an ML-KEM key");
+        return RET_OSSL_ERR;
+    }
+
+    ctx->op = op;
+    return p11prov_mlkem_set_ctx_params(vctx, params);
+}
+
+static int p11prov_mlkem_encapsulate_init(void *vctx, void *vkey,
+                                          const OSSL_PARAM params[])
+{
+    P11PROV_OBJ *key = vkey;
+
+    if (p11prov_obj_get_class(key) != CKO_PUBLIC_KEY) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_KEY);
```

Comment:
Would a `PROV_R_NOT_A_PUBLIC_KEY` be a better error here? Same in the function below, `PROV_R_NOT_A_PRIVATE_KEY`?

Reply:
sure